File: compare.py
import json
import re

def normalize_name(name):
    name = name.lower()
    for b in ['realme', 'samsung', 'oppo', 'vivo', 'hmd', 'nokia', 'itel', 'lava', 'motorola', 'infinix', 'tecno']:
        name = name.replace(b, '')
    
    # Remove things like 4+64, 8+128
    name = re.sub(r'\d+\s*\+\s*\d+', '', name)
    # Remove colors in brackets
    name = re.sub(r'\(.*?\)', '', name)
    # Remove extra spaces
    name = re.sub(r'\s+', ' ', name).strip()
    return name

# Load TS data
ts_content = open('d:/downloadss/shreeshyam mobiles/products-data.ts', 'r', encoding='utf-8').read()
web_products = []
for match in re.finditer(r'name:\s*"([^"]+)",\s*brand:\s*"([^"]+)",\s*price:\s*(\d+),\s*ram:\s*"([^"]+)",\s*storage:\s*"([^"]+)"', ts_content):
    web_products.append({
        'name': match.group(1).lower(),
        'norm_name': normalize_name(match.group(1)),
        'brand': match.group(2).lower(),
        'price': int(match.group(3)),
        'ram': match.group(4).lower().replace('gb', ''),
        'storage': match.group(5).lower().replace('gb', '')
    })

# Load JSON data
json_content = open('d:/downloadss/shreeshyam mobiles/All_Brands_Price_List (1).json', 'r', encoding='utf-8').read()
json_data = json.loads(json_content)

json_flat = []
for brand, products in json_data['brands'].items():
    for p in products:
        json_flat.append({
            'name': p['product_name'].lower(),
            'norm_name': normalize_name(p['product_name']),
            'brand': brand.lower(),
            'price': p['selling_price'],
            'ram': str(p['ram_gb']).lower(),
            'storage': str(p['storage_gb']).lower()
        })

match_count = 0
mismatch_count = 0
not_found = 0

output = "# Price Comparison Report\n\n"
output += "| JSON Product | RAM/Storage | JSON Price | Web Price | Status |\n"
output += "|---|---|---|---|---|\n"

for jp in json_flat:
    # Find matching web product
    # Try exact or partial name match with exact ram/storage
    possible_matches = []
    jp_words = jp['norm_name'].split()
    
    for wp in web_products:
        if wp['brand'] != jp['brand']: continue
        
        # Check ram/storage (handle nulls and 1tb = 1024)
        if jp['ram'] != 'null' and wp['ram'] != 'null' and jp['ram'] != wp['ram']: continue
        
        w_storage = '1024' if wp['storage'] == '1tb' else wp['storage']
        if jp['storage'] != 'null' and wp['storage'] != 'null' and jp['storage'] != w_storage: continue
        
        # Word inclusion
        matches = True
        for w in jp_words:
            if len(w) > 2 and w not in wp['norm_name']:
                matches = False
                break
                
        if matches:
            possible_matches.append(wp)
            
    if len(possible_matches) > 0:
        match = possible_matches[0]
        if match['price'] == jp['price']:
            match_count += 1
        else:
            mismatch_count += 1
            output += f"| {jp['name']} | {jp['ram']}/{jp['storage']} | ₹{jp['price']} | ₹{match['price']} | ❌ MISMATCH |\n"
    else:
        not_found += 1

summary = f"## Summary\n"
summary += f"- **Total in JSON:** {len(json_flat)}\n"
summary += f"- **Total in Website:** {len(web_products)}\n"
summary += f"- **Perfect Matches:** {match_count}\n"
summary += f"- **Price Mismatches:** {mismatch_count}\n"
summary += f"- **Could not match:** {not_found}\n\n"

open('d:/downloadss/shreeshyam mobiles/price-comparison.md', 'w', encoding='utf-8').write(summary + output)
print("Comparison done. Check price-comparison.md")

File: compare.js
const fs = require('fs');

// Read website data
const tsContent = fs.readFileSync('d:/downloadss/shreeshyam mobiles/products-data.ts', 'utf8');
const regex = /name:\s*"([^"]+)",\s*brand:\s*"([^"]+)",\s*price:\s*(\d+),\s*ram:\s*"([^"]+)",\s*storage:\s*"([^"]+)"/g;
let match;
const webProducts = [];
while ((match = regex.exec(tsContent)) !== null) {
  webProducts.push({ 
    name: match[1].toLowerCase(), 
    brand: match[2].toLowerCase(), 
    price: parseInt(match[3]), 
    ram: match[4].replace(/gb/i, ''), 
    storage: match[5].replace(/gb/i, '') 
  });
}

// Read JSON data
const jsonContent = fs.readFileSync('d:/downloadss/shreeshyam mobiles/All_Brands_Price_List (1).json', 'utf8');
const jsonData = JSON.parse(jsonContent);

let matchCount = 0;
let mismatchCount = 0;
let notFoundInWeb = 0;
let notFoundInJson = 0;
let output = '';

output += '# Price Comparison Report\n\n';

// Normalize helper
const normalizeName = (name) => {
  return name.toLowerCase()
    .replace(/realme\s*/, '')
    .replace(/samsung\s*/, '')
    .replace(/oppo\s*/, '')
    .replace(/vivo\s*/, '')
    .replace(/\s+/g, ' ')
    .replace(/(\d+)\s*\+\s*(\d+)/, '') // remove ram+storage from name if any
    .replace(/\(.*?\)/g, '') // remove colors in brackets
    .trim();
};

const jsonFlat = [];
for (const brand in jsonData.brands) {
  jsonData.brands[brand].forEach(p => {
    jsonFlat.push({
      originalName: p.product_name,
      name: normalizeName(p.product_name),
      brand: brand.toLowerCase(),
      ram: String(p.ram_gb).toLowerCase(),
      storage: String(p.storage_gb).toLowerCase(),
      price: p.selling_price
    });
  });
}

const webFlat = webProducts.map(p => ({
  originalName: p.name,
  name: normalizeName(p.name),
  brand: p.brand,
  ram: p.ram.toLowerCase(),
  storage: p.storage.toLowerCase(),
  price: p.price
}));

// Compare JSON -> Web
output += '## Mismatches (JSON -> Website)\n\n';
output += '| JSON Product | RAM/Storage | JSON Price | Web Price | Status |\n';
output += '|---|---|---|---|---|\n';

jsonFlat.forEach(jp => {
  // Try to find matching product in web
  // We'll use a fuzzy match on name and exact match on RAM/Storage
  
  // Clean names for matching
  const jNameParts = jp.name.split(' ');
  
  const possibleMatches = webFlat.filter(wp => {
    // Check if RAM and Storage match
    if (wp.ram !== jp.ram && wp.ram !== 'null' && jp.ram !== 'null') return false;
    
    // Some products have storage like "1TB" which is 1024GB, let's keep it simple for now
    let wStorage = wp.storage;
    if (wStorage === '1tb') wStorage = '1024';
    
    if (wStorage !== jp.storage && wStorage !== 'null' && jp.storage !== 'null') return false;
    
    // Check brand
    if (wp.brand !== jp.brand) return false;
    
    // Check name inclusion
    let matches = true;
    for (let part of jNameParts) {
      if (part.length > 2 && !wp.name.includes(part)) {
        matches = false;
        break;
      }
    }
    return matches;
  });
  
  if (possibleMatches.length > 0) {
    // Check price of the first match
    const match = possibleMatches[0];
    if (match.price === jp.price) {
      matchCount++;
    } else {
      mismatchCount++;
      output += `| ${jp.originalName} | ${jp.ram}/${jp.storage} | ₹${jp.price} | ₹${match.price} | ❌ MISMATCH |\n`;
    }
  } else {
    notFoundInWeb++;
    // Too many "Not Found" to list them all, we'll just log mismatches
  }
});

output = `## Summary
- **Total in JSON:** ${jsonFlat.length}
- **Total in Website:** ${webFlat.length}
- **Perfect Matches:** ${matchCount}
- **Price Mismatches:** ${mismatchCount}
- **Could not match (JSON to Web):** ${notFoundInWeb}

*Note: "Could not match" means the script couldn't reliably link the JSON product to a Website product based on name, RAM, and storage.*

` + output;

fs.writeFileSync('d:/downloadss/shreeshyam mobiles/price-comparison.md', output, 'utf8');
console.log('Comparison complete. Summary:');
console.log(`Matches: ${matchCount}, Mismatches: ${mismatchCount}, Not Found: ${notFoundInWeb}`);
